File: implementations/ios-sdk/OptimizationApp/Components/NestedContentEntryView.swift
import ContentfulOptimization
import SwiftUI

struct NestedContentEntryView: View {
    let entry: [String: Any]

    private var entryId: String {
        let sys = entry["sys"] as? [String: Any]
        return sys?["id"] as? String ?? ""
    }

    private var nestedEntries: [[String: Any]] {
        let fields = entry["fields"] as? [String: Any]
        guard let nestedArray = fields?["nested"] as? [Any] else { return [] }
        return nestedArray.compactMap { $0 as? [String: Any] }.filter { item in
            guard let sys = item["sys"] as? [String: Any] else { return false }
            return sys["id"] != nil
        }
    }

    var body: some View {
        VStack(alignment: .leading) {
            OptimizedEntry(
                entry: entry,
                accessibilityIdentifier: "content-entry-\(entryId)"
            ) { resolvedEntry in
                NestedEntryText(entry: resolvedEntry)
            }

            ForEach(0..<nestedEntries.count, id: \.self) { index in
                NestedContentEntryView(entry: nestedEntries[index])
            }
        }
    }
}

private struct NestedEntryText: View {
    let entry: [String: Any]

    private var entryId: String {
        let sys = entry["sys"] as? [String: Any]
        return sys?["id"] as? String ?? ""
    }

    private var text: String {
        let fields = entry["fields"] as? [String: Any]
        return fields?["text"] as? String ?? "No content"
    }

    var body: some View {
        VStack(alignment: .leading) {
            Text(text)
            Text("[Entry: \(entryId)]")
        }
        .padding()
        .accessibilityElement(children: .ignore)
        .accessibilityLabel("\(text) [Entry: \(entryId)]")
    }
}

File: implementations/android-sdk/app/src/main/kotlin/com/contentful/optimization/app/components/NestedContentEntryView.kt
package com.contentful.optimization.app.components

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import com.contentful.optimization.compose.OptimizedEntry

@Composable
fun NestedContentEntryView(entry: Map<String, Any>) {
    val entryId = entryId(entry)

    @Suppress("UNCHECKED_CAST")
    val fields = entry["fields"] as? Map<String, Any>
    val nestedArray = fields?.get("nested") as? List<*> ?: emptyList<Any>()

    @Suppress("UNCHECKED_CAST")
    val nestedEntries = nestedArray.filterIsInstance<Map<String, Any>>().filter { item ->
        val sys = item["sys"] as? Map<String, Any>
        sys?.get("id") != null
    }

    Column {
        OptimizedEntry(
            entry = entry,
            accessibilityIdentifier = "content-entry-$entryId",
        ) { resolvedEntry ->
            NestedEntryText(entry = resolvedEntry)
        }

        nestedEntries.forEach { nestedEntry ->
            NestedContentEntryView(entry = nestedEntry)
        }
    }
}

@Composable
private fun NestedEntryText(entry: Map<String, Any>) {
    val id = entryId(entry)
    @Suppress("UNCHECKED_CAST")
    val fields = entry["fields"] as? Map<String, Any>
    val text = fields?.get("text") as? String ?: "No content"

    Column(
        modifier = Modifier
            .padding(16.dp)
            .testTag("entry-text-$id")
            .semantics(mergeDescendants = true) {
                contentDescription = "$text [Entry: $id]"
            },
    ) {
        Text(text)
        Text("[Entry: $id]")
    }
}

@Suppress("UNCHECKED_CAST")
fun isNestedContent(entry: Map<String, Any>): Boolean {
    val sys = entry["sys"] as? Map<String, Any> ?: return false
    val contentType = sys["contentType"] as? Map<String, Any> ?: return false
    val innerSys = contentType["sys"] as? Map<String, Any> ?: return false
    val id = innerSys["id"] as? String ?: return false
    return id == "nestedContent"
}
